haproxy/haproxy.go

```diff
@@ -183,14 +189,78 @@ func (h *HAproxy) WriteConfig(state *catalog.ServicesState, output io.Writer) er
 	return nil
 }
 
-// Execute a command and bubble up the error
+// notifySignals swallows a bunch of signals that get sent to us when running into
+// an error from HAproxy. If we didn't swallow these, the process would potentially
+// stop when the signals are propagated by the sub-shell.
+func (h *HAproxy) swallowSignals() {
+	// from HAproxy which propagate.
+	sigChan := make(chan os.Signal, 10)
+
+	// Used to stop the goroutine
+	h.sigStopChan = make(chan struct{})
+
+	go func() {
+		for {
+			select {
+			case <-sigChan:
+				// swallow signal
+			case <-h.sigStopChan:
+				break
+			}
+		}
+	}()
+
+	signal.Notify(sigChan, syscall.SIGSTOP, syscall.SIGTSTP, syscall.SIGTTIN, syscall.SIGTTOU)
+}
+
+// ResetSignals unhooks our signal handler from the signals the sub-commands
+// initiate. This is potentially destructive if other places in the program have
+// hooked to the same signals! Affected signals are SIGSTOP, SIGTSTP, SIGTTIN, SIGTTOU.
+func (h *HAproxy) ResetSignals() {
+	h.sigLock.Lock()
+	signal.Reset(syscall.SIGSTOP, syscall.SIGTSTP, syscall.SIGTTIN, syscall.SIGTTOU)
+	select {
+	case h.sigStopChan <- struct{}{}: // nothing
+	default:
+	}
+
+	h.sigLock.Unlock()
+}
+
+// Execute a command and bubble up the error. Includes locking behavior which means
+// that only one of these can be running at once.
 func (h *HAproxy) run(command string) error {
-	out, err := exec.Command("/bin/bash", "-c", command).CombinedOutput()
+
+	cmd := exec.Command("/bin/bash", "-c", command)
+	stdout := &bytes.Buffer{}
+	stderr := &bytes.Buffer{}
+	cmd.Stdout = stdout
+	cmd.Stderr = stderr
+
+	// The end effect of this signal handling requirement is that we can only run _one_
+	// command at a time. This is totally fine for HAproxy.
+	h.sigLock.Lock()
+	defer h.sigLock.Unlock()
+
+	if !h.signalsHandled {
+		log.Info("Setting up signal handlers")
+		h.swallowSignals()
+		h.signalsHandled = true
+	}
+
+	err := cmd.Start()
+
 	if err != nil {
-		return fmt.Errorf("Error running '%s': %s\n%s", command, err.Error(), out)
+		return fmt.Errorf("Unable to start '%s': %s", command, err)
 	}
 
-	return nil
+	err = cmd.Wait()
```

Comment:
Looks like it: https://golang.org/src/os/exec/exec.go?s=7708:7733#L266 . I had originally implemented a timeout here, but since this has been in production for a long time and we've never seen that happen, I removed it.